### ISSUE TITLE
Benchmarks for pyfixest

### ISSUE THREAD

**Issue by aeturrell** [s3alfisc/pyfixest#202]
This one might be happening anyway given I saw #181 by [Grant](https://github.com/grantmcdermott), but it would be quite useful to know how fast pyfixest is relative to the other packages out there.

The easiest way to achieve it would probably be to convince [Laurent](https://github.com/lrberge/fixest/) to add it to the [benchmarks he runs for the docs for fixest](https://lrberge.github.io/fixest/#benchmarking).

I'm interested because I'm wondering what parts of the code-base of pyfixest could be sped up by a Rust backend. To be clear, I've never used Rust, but there have been a lot of successes in [using it as a Python back-end](https://developers.redhat.com/blog/2017/11/16/speed-python-using-rust#what_is_rust_) recently (see [Ruff](https://docs.astral.sh/ruff/) and [Polars](https://duckdblabs.github.io/db-benchmark/)), so seems worth thinking about. Of course it depends how quickly #66 proceeds.

**Comment by s3alfisc** [comment]
Yes, I should absolutely tackle benchmarking more systematically. My gut feeling is there there is still some scope for improvement in the preprocessing steps (besides the main demeaning algo). I just did some small checks using fixest's example data sets in #203 and was surprised (shocked!) how slow `pyfixest` was in comparison. The culprit were clustered SEs with 50K clusters - turns out I need to optimize CRV performance #204!

**Comment by grantmcdermott** [comment]
Quick 2c on this:

My own benchmarks indicate that pyfixest is about 2-4 times slower than r-fixest for standard operations. That's a huge improvement on where it was even a month or so ago and, once again, massive props to Alex for all his work.

However, something that makes me very uncomfortable is the divergence in the underlying codebases... particularly w.r.t to the low-level computation engines. It's one thing to mimic the syntax and superficial behaviour, and both R and Python come with their own idiosyncrasies that would have ameliorated no matter what. But I've seen multi-language projects fall into disarray often enough, because of divergence at the C/C++/etc. level. Contrast this with projects that deliberately manage low-level dependence well (DuckDB, arrow, etc.)

(As an aside, I'm concerned that polars is that slipping into the former category. Their development cycle still has the feel of the Wild West and breaking changes--esp. at the Rust level--are very poorly documented.) 

My personal opinion is that the longevity of this project---as a fixest clone---heeds caution against a new a Rust backend, or even rely further on numba (although that's probably a bridge too far at this point). Much better to work towards a common C++ API and help Laurent on development side there. To clarify, I don't mean to stop you from going further down a Rust/etc if it works from you. But IMO you risk disservice to both implementations by continuing under the "fixest" moniker as the codebases diverge even further.

**Comment by s3alfisc** [comment]
Thanks for your feedback, @grantmcdermott!

> My personal opinion is that the longevity of this project---as a fixest clone---heeds caution against a new a Rust backend, or even rely further on numba (although that's probably a bridge too far at this point). 

I agree with all you say here. If possible, I'd prefer to call a common fixest C++ API. For now, I will stay with `numba` and move to the C++ API if / whenever it becomes available. It's not that I would not be curious to play around with rust @aeturrell , but the fixest implementation is quite complex & @lrberge suggested I should not try it to implement some of its more specialised features (mostly interactions of continuous variables with fixed effects) by myself if I wanted to stay sane 😄 . 

Though I am also a little bit worried that what @grantmcdermott suggests - to take `duckdb` or `arrow` as role models for asuccessful multi language project - might be **a lot** of work for Laurent, me, or anyone involved. `fixest` is not just the fast demeaning algorithm, but in fact a set of optimized algorithms (i.e. detecting multicollinearity, fast CRV inference, and I think it even inverts matrices by hand and in parallel when solving the OLS equation) via custom c++ implementations. But then as with everything, I suppose that one could start with the most important ones and then gradually build from there. 😃

**Comment by s3alfisc** [comment]
Hi all, I finally got around to run some of the `fixest` benchmarks, and here are the results: `pyfixest` seems on par with `lfe`, which I think is already pretty decent =) But of course it's slower than `fixest`. I'll post a link to the branch with code tomorrow. 
![image](https://github.com/s3alfisc/pyfixest/assets/19531450/95081dc5-4e8d-4e1d-8af8-4458a86766ff)

(@styfenschaer as a lot of the good performance is coming from your work - thanks again for it!)

**Comment by s3alfisc** [comment]
You can find the benchmarks for OLS and Poisson + code in [this branch](https://github.com/s3alfisc/pyfixest/tree/crv-benchmarks).

**Comment by s3alfisc** [comment]
Closed with #219 .

**Comment by s3alfisc** [comment]
Yes, putting this together has actually been already super useful - if I had not done it, it probably would have taken me much longer to realize the CRV1 performance bottleneck =)  
> I think the performance is particularly impressive given the numbers of lines of code (although I realise this doesn't necessarily correlate with effort!).

I'm also super happy with the performance! Probs to @styfenschaer for helping me out with all things `numba`, which made quite a difference. Do you have a suggestion what I should focus on next @aeturrell? I'll have some free days starting from mid December and am trying to collect a set of additional features / improvements [here](https://github.com/s3alfisc/pyfixest/issues/218). So it's going to be more code in the future =)

**Comment by aeturrell** [comment]
I have some suggestions informed by what I've got in Coding for Economists and by what I want to do in my teaching, which is make regressions fit smoothly into a workflow. But it's *definitely* worth canvassing others as these may not be what people need at the coal face, and I'm not sure how big a deal any of them would be to implement.

1. Generalised linear models. Not sure if this is totally out of scope. In [current draft](https://github.com/aeturrell/coding-for-economists/blob/main/econmt-regression.ipynb) of CfE's regression page I drop back into **statsmodels** for this. Which is fine, and it's an excellent package, but for pedagogical reasons it can be better to cover only one package.
2. Further regression stats. **statsmodels** [produces a lot](https://www.statsmodels.org/stable/index.html), perhaps too many. But, for example, F-stats would be really useful.
3. Output tables. The easier it is to produce publication-ready tables, the more I think it would help users adopt the package. In **fixest**, [details such as no. obs, R squared, etc, etc](https://lrberge.github.io/fixest/articles/exporting_tables.html), are included in exported tables that include multiple models. In practice, this is what is needed in a publication. At the moment, `etable` (**pyfixest** version 0.10.7) doesn't have that detail. Maybe this should be first! Although I realise it may be more complex to implement.

**Comment by s3alfisc** [comment]
Thanks! These suggestions make a lot of sense and there is good overlap with what I had already in mind. 
I will start with etable improvements & regression diagnostics, as these might be the lowest hanging fruits. So far adding more GLM support was not really on my roadmap, but now that you mention it, I will give it a thought! I suppose that logistic regression would be the most relevant GLM? Would you prioritize it  over additional difference-in-differences methods, as e.g. the Sun-Abraham estimator?

**Comment by s3alfisc** [comment]
Just to give some feedback (I realized I hadn't yet responded - sorry!). I'll try to push for a new release before Christmas, with a methods for Wald/F Tests, some variants of the local projections difference in differences estimator, and R2 values in etable. And if I can get this done in time, I might take a look at logit between the years =)

**Comment by s3alfisc** [comment]
Btw the CfE's draft looks awesome, @aeturrell!
